clients/spark/core/src/main/scala/io/treeverse/clients/LakeFSInputFormat.scala

```diff
@@ -1,74 +1,145 @@
+import java.io.DataInput
+import java.io.DataOutput
+import java.io.File
+import java.net.URI
 import scala.collection.JavaConverters._
+import scala.collection.mutable.ListBuffer
+import com.google.protobuf.ByteString
+import org.slf4j.{Logger, LoggerFactory}
 
-class GravelerSplit(var range: RangeData, var path: Path) extends InputSplit with Writable {
-  def this() = this(null, null)
+object GravelerSplit {
+  val logger: Logger = LoggerFactory.getLogger(getClass.toString)
+}
+class GravelerSplit(var range: RangeData, var path: Path, var rangeID: String, var byteSize: Long)
+    extends InputSplit
+    with Writable {
+  import GravelerSplit._
+  def this() = this(null, null, null, 0L)
 
   override def write(out: DataOutput): Unit = {
-    val encodedRange = range.toByteArray
-    out.writeInt(encodedRange.length)
-    out.write(encodedRange)
+    out.writeLong(byteSize)
+    out.writeInt(rangeID.length)
+    out.writeChars(rangeID)
+    if (range != null) {
+      val encodedRange = range.toByteArray
+      out.writeInt(encodedRange.length)
+      out.write(encodedRange)
+    } else {
+      out.writeInt(0)
+    }
     val p = path.toString
     out.writeInt(p.length)
     out.writeChars(p)
   }
 
   override def readFields(in: DataInput): Unit = {
+    byteSize = in.readLong()
+    val rangeIDLength = in.readInt()
+
+    val rangeIDBuilder = new StringBuilder
+    for (_ <- 1 to rangeIDLength) {
+      rangeIDBuilder += in.readChar()
+    }
+    rangeID = rangeIDBuilder.toString
     val encodedRangeLength = in.readInt()
-    val encodedRange = new Array[Byte](encodedRangeLength)
-    in.readFully(encodedRange)
-    range = RangeData.parseFrom(encodedRange)
+    if (encodedRangeLength == 0) {
+      range = null
+    } else {
+      val encodedRange = new Array[Byte](encodedRangeLength)
+      in.readFully(encodedRange)
+      range = RangeData.parseFrom(encodedRange)
+    }
     val pathLength = in.readInt()
     val p = new StringBuilder
     for (_ <- 1 to pathLength) {
       p += in.readChar()
     }
     path = new Path(p.result)
+    logger.debug(s"Read split $this")
   }
-
-  override def getLength: Long = range.estimatedSize
+  override def getLength: Long = byteSize
 
   override def getLocations: Array[String] = Array.empty[String]
 
   override def getLocationInfo: Array[SplitLocationInfo] =
     Array.empty[SplitLocationInfo]
+
+  override def toString: String = {
+    val sb = new StringBuilder
+    sb.append("GravelerSplit: ")
+    sb.append("rangeID: ")
+    sb.append(new String(rangeID))
+    sb.append("\n")
+    sb.append("path: ")
+    sb.append(path)
+    sb.append("\n")
+    sb.append("range: ")
+    sb.append(range)
+    sb.toString
+  }
 }
 
 class WithIdentifier[Proto <: GeneratedMessage with scalapb.Message[Proto]](
     val id: Array[Byte],
-    val message: Proto
+    val message: Proto,
+    val rangeID: String
 ) {}
 
+object EntryRecordReader {
+  val logger: Logger = LoggerFactory.getLogger(getClass.toString)
```

Comment:
By putting it on a companion object, you get a weird class name:
```scala
scala> object X { val className = getClass.toString }
object X
scala> X.className
val res0: String = class X$

scala> class X { val className = getClass.toString }
scala> (new X).className
val res2: String = class X
```

Why not just put it inside the class?

Reply:
If it's inside the class, it's created together with every instance. Thoughts?

Reply:
I am not sure whether LoggerFactory actually creates a new instance.  And there are not that many RecordReaders in existence.  Your way gives a strange classname, so I would rather avoid it.  (And I don't know how to access the Class object of the class from the companion object, sorry.)